src/main/java/DirectoryEncryptor/ASyncDirectoryEncryptor.java

```diff
@@ -0,0 +1,100 @@
+package DirectoryEncryptor;
+
+import EncryptionAlgorithms.EncryptionAlgorithm;
+
+import EventsLogger.Events;
+import Exceptions.InvalidKeyException;
+import Keys.Key;
+import Utils.FileOperations;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+public class ASyncDirectoryEncryptor extends DirectoryProcessor {
+    private String pathToDir;
+    private FileOperations directoryOperations;
+
+    private FileOperations keyFile;
+    private EncryptionAlgorithm encryptionAlgorithm;
+    private List<Thread> encryptThreadList = new ArrayList<>();
+    private List<Thread> decryptThreadList = new ArrayList<>();
```

Comment:
Use ScheduledExecutorService instead list of threads

Reply:
Attention to critical sections (file write etc.) - Use synchronize (for example using lock or synchronize block) in those code sections